lib/codecs/src/encoding/format/cef.rs

```diff
@@ -0,0 +1,520 @@
+    },
+    #[snafu(display(
+        r#"LogEvent CEF severity must be a number from 0 to {}: actual {}"#,
+        max_value,
+        actual_value
+    ))]
+    SeverityMaxValue { max_value: u8, actual_value: u8 },
+    #[snafu(display(r#"LogEvent CEF severity must be a number: {}"#, error))]
+    SeverityNumberType { error: ParseIntError },
+    #[snafu(display(r#"LogEvent extension keys can only contain ascii alphabetical characters: invalid key "{}""#, key))]
+    ExtensionNonASCIIKey { key: String },
+}
+
+/// Config used to build a `CefSerializer`.
+#[crate::configurable_component]
+#[derive(Debug, Clone)]
+pub struct CefSerializerConfig {
+    /// The CEF Serializer Options.
+    pub cef: CefSerializerOptions,
+}
+
+impl CefSerializerConfig {
+    /// Creates a new `CefSerializerConfig`.
+    pub const fn new(cef: CefSerializerOptions) -> Self {
+        Self { cef }
+    }
+
+    /// Build the `CefSerializer` from this configuration.
+    pub fn build(&self) -> Result<CefSerializer, BuildError> {
+        let device_vendor = escape_header(self.cef.device_vendor.as_str());
+        if device_vendor.len() > DEVICE_VENDOR_MAX_LENGTH {
+            return ExceededLengthSnafu {
+                field: device_vendor.clone(),
+                max_length: DEVICE_VENDOR_MAX_LENGTH,
+                actual_length: device_vendor.len(),
+            }
+            .fail()
+            .map_err(|e| e.to_string().into());
+        };
+
+        let device_product = escape_header(self.cef.device_product.as_str());
+        if device_product.len() > DEVICE_PRODUCT_MAX_LENGTH {
+            return ExceededLengthSnafu {
+                field: device_product.clone(),
+                max_length: DEVICE_PRODUCT_MAX_LENGTH,
+                actual_length: device_product.len(),
+            }
+            .fail()
+            .map_err(|e| e.to_string().into());
+        };
+
+        let device_version = escape_header(self.cef.device_version.as_str());
+        if device_version.len() > DEVICE_VERSION_MAX_LENGTH {
+            return ExceededLengthSnafu {
+                field: device_version.clone(),
+                max_length: DEVICE_VERSION_MAX_LENGTH,
+                actual_length: device_version.len(),
+            }
+            .fail()
+            .map_err(|e| e.to_string().into());
+        };
+
+        let device_event_class_id = escape_header(self.cef.device_event_class_id.as_str());
+        if device_event_class_id.len() > DEVICE_EVENT_CLASS_ID_MAX_LENGTH {
+            return ExceededLengthSnafu {
+                field: device_event_class_id.clone(),
+                max_length: DEVICE_EVENT_CLASS_ID_MAX_LENGTH,
+                actual_length: device_event_class_id.len(),
+            }
+            .fail()
+            .map_err(|e| e.to_string().into());
+        };
+
+        for key in self.cef.extensions.keys() {
+            if !key.chars().all(|c| c.is_ascii_alphabetic()) {
+                // TODO: Output all invalid keys
```

Comment:
The following code should address this TODO:
```
    // Check for non-ASCII keys in extensions
    let invalid_keys: Vec<String> = self.cef.extensions.keys()
        .filter(|key| !key.chars().all(|c| c.is_ascii_alphabetic()))
        .cloned()
        .collect();
    
    if !invalid_keys.is_empty() {
        return ExtensionNonASCIIKeySnafu { key: invalid_keys.join(", ") }.fail();
    }
```

Reply:
Took this one, thanks